### PR TITLE
Progressive device tree scanning.

### PR DESCRIPTION
**Pull request by MichalBlk**
Scan the device tree multiple times. Each scan is identified be a pass number and each device driver is assigned a pass number.
At each scan, drivers can be attached to devices. Driver can be probed and attached only if the current system-wide pass number
matches the driver's pass number. The first pass is devoted to detect basic hardware resources like timers and buses. All remaining devices are processed during the second (FTTB also last) pass.

### CODE REVIEW

File: include/sys/device.h [hunk 1]
```diff
@@ -29,13 +29,23 @@ typedef enum {
...
+} pass_num_t;
+
 struct driver {
   const char *desc;            /* short driver description */
   size_t size;                 /* device->state object size */
   d_probe_t probe;             /* probe for specific device(s) */
   d_attach_t attach;           /* attach device to system */
   d_detach_t detach;           /* detach device from system */
   void *interfaces[DIF_COUNT]; /* pointers to device methods structures */
+  pass_num_t pass;             /* device tree pass number */
```
Comment (cahirwpz) [inline include/sys/device.h:48]: Could you move it just after `size` - here and in every definition?

File: include/sys/bus.h [hunk 2]
```diff
@@ -210,6 +210,9 @@ static inline void bus_release_resource(device_t *dev, res_type_t type,
   BUS_METHODS(idev->parent).release_resource(idev, type, r);
 }
 
+/* System-wide current pass number. */
+extern pass_num_t current_pass;
```
Comment (cahirwpz) [inline include/sys/bus.h:214]: I think in current use scenario you can get rid of it.

File: sys/aarch64/rootdev.c [hunk 3]
```diff
@@ -302,21 +304,27 @@ static bus_methods_t rootdev_bus_if = {
...
-  .state = &(rootdev_t){},
-  .devclass = &DEVCLASS(root),
-};
-
-DEVCLASS_ENTRY(root, rootdev);
+DEVCLASS_CREATE(root);
 
-void init_devices(void) {
-  device_attach(&rootdev);
+void init_devices(pass_num_t pass) {
```
Comment (cahirwpz) [inline sys/aarch64/rootdev.c:318]: Please note that first pass is performed when `rootdev` is `NULL`. Other passes when `rootdev` is already present.

File: sys/aarch64/trap.c [hunk 4]
```diff
@@ -18,7 +18,8 @@ static __noreturn void kernel_oops(ctx_t *ctx) {
   panic();
 }
 
-static void syscall_handler(register_t code, ctx_t *ctx, syscall_result_t *result) {
+static void syscall_handler(register_t code, ctx_t *ctx,
```
Comment (cahirwpz) [inline sys/aarch64/trap.c:21]: Random change?

File: sys/kern/bus.c [hunk 5]
```diff
@@ -117,26 +117,41 @@ void bus_deactivate_resource(device_t *dev, res_type_t type, resource_t *r) {
   rman_deactivate_resource(r);
 }
 
+pass_num_t current_pass;
```
Comment (cahirwpz) [inline sys/kern/bus.c:120]: Please make it `static`.

File: sys/kern/main.c [hunk 6]
```diff
@@ -99,8 +102,8 @@ __noreturn void kernel_init(void) {
   /* Mount filesystems (including devfs). */
   mount_fs();
 
-  /* First (FTTB also last) stage of device init. */
-  init_devices();
+  /* First stage of device init. */
+  init_devices(FIRST_PASS);
```
Comment (cahirwpz) [inline sys/kern/main.c:106]: It'd be great if you described what kind of drivers should be initialized by `FIRST_PASS`. Our goal was to start those that do not require `devfs`, but provide first timer and console output.

**Review by cahirwpz — COMMENTED** [review]
I like the simplicity, but `FreeBSD` does it in different way. Can you look into their sources and figure out why they need such a complex solution? (they keep `pass` field in `driverlink`)

File: include/sys/device.h (hunk 1 above)
Comment (MichalBlk) [reply include/sys/device.h]: Done.

File: include/sys/bus.h (hunk 2 above)
Comment (MichalBlk) [reply include/sys/bus.h]: Done.

File: sys/aarch64/trap.c (hunk 4 above)
Comment (MichalBlk) [reply sys/aarch64/trap.c]: This is due to make format for aarch64.

File: sys/kern/bus.c (hunk 5 above)
Comment (MichalBlk) [reply sys/kern/bus.c]: Done.

File: sys/kern/main.c (hunk 6 above)
Comment (MichalBlk) [reply sys/kern/main.c:107]: Done.

File: sys/aarch64/rootdev.c (hunk 3 above)
Comment (MichalBlk) [reply sys/aarch64/rootdev.c]: Done.

**Review by cahirwpz — APPROVED** [review]
LGTM

**Pull request merged**